pkg/controllers/updaterun/controller.go

```diff
@@ -0,0 +1,158 @@
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	placementv1alpha1 "go.goms.io/fleet/apis/placement/v1alpha1"
+)
+
+const (
+	// stageUpdateRunDeletionWaitTime is the time to wait before checking whether the stage update run is completed before deletion.
+	stageUpdateRunDeletionWaitTime = 60 * time.Second
+)
+
+// Reconciler reconciles a ClusterStagedUpdateRun object.
+type Reconciler struct {
+	client.Client
+	recorder record.EventRecorder
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req runtime.Request) (runtime.Result, error) {
+	startTime := time.Now()
+	klog.V(2).InfoS("ClusterStagedUpdateRun reconciliation starts", "clusterStagedUpdateRun", req.NamespacedName)
+	defer func() {
+		latency := time.Since(startTime).Milliseconds()
+		klog.V(2).InfoS("ClusterStagedUpdateRun reconciliation ends", "clusterStagedUpdateRun", req.NamespacedName, "latency", latency)
+	}()
+
+	var updateRun placementv1alpha1.ClusterStagedUpdateRun
+	if err := r.Client.Get(ctx, req.NamespacedName, &updateRun); err != nil {
+		klog.ErrorS(err, "Failed to get clusterStagedUpdateRun object", "clusterStagedUpdateRun", req.Name)
+		return runtime.Result{}, client.IgnoreNotFound(err)
+	}
+	runObjRef := klog.KObj(&updateRun)
+
+	// Handle the deletion of the clusterStagedUpdateRun.
+	if !updateRun.DeletionTimestamp.IsZero() {
+		klog.V(2).InfoS("The clusterStagedUpdateRun is being deleted", "clusterStagedUpdateRun", runObjRef)
+		deleted, waitTime, err := r.handleDelete(ctx, updateRun.DeepCopy())
+		if err != nil {
+			return runtime.Result{}, err
+		}
+		if deleted {
+			return runtime.Result{}, nil
+		}
+		return runtime.Result{RequeueAfter: waitTime}, nil
+	}
+
+	// Add the finalizer to the clusterStagedUpdateRun.
+	if err := r.ensureFinalizer(ctx, &updateRun); err != nil {
+		klog.ErrorS(err, "Failed to add the finalizer to the clusterStagedUpdateRun", "clusterStagedUpdateRun", runObjRef)
+		return runtime.Result{}, err
+	}
+
+	// TODO(wantjian): reconcile the clusterStagedUpdateRun.
+	return runtime.Result{}, nil
+}
+
+// handleDelete handles the deletion of the clusterStagedUpdateRun object.
+// We need to wait for the update run to stop before deleting the clusterStagedUpdateRun object.
+// We will delete all the dependent resources, including approvalRequest objects, of the clusterStagedUpdateRun object.
+func (r *Reconciler) handleDelete(ctx context.Context, updateRun *placementv1alpha1.ClusterStagedUpdateRun) (bool, time.Duration, error) {
+	runObjRef := klog.KObj(updateRun)
+	// check if the update run is still running.
+	// the update run is considered as running if started condition is true while finished condition is not true or false.
```

Comment:
why wait for the update run to complete? It can be stuck (with no approval or one application not working on one cluster)? How can we delete it then?

Reply:
removed this check.